exercises/edward_brown/ex01_basics/include/TakeWhile.h

```diff
@@ -0,0 +1,31 @@
+#ifndef EX1_TAKE_WHILE
+#define EX1_TAKE_WHILE
+#include <sstream>
+#include <utility>
+
+/** Builds a substring s of the range [Begin, End) such that s contains
+ *  the characters between the start of the range and the character c where
+ *  c is the first occurence of a character such that ShouldTake(c) is false.
+ *  If no such character exists then the substring s will we equal to the string
+ *  of all characters in the range [Begin, End)
+ *
+ *  @param Begin The const_iterator marking the begining of the range to be
+ *  searched for c.
+ *  @param End The const_iterator marking the end of the range to be
+ *  searched for c.
+ *  @param ShouldTake A callable object which returns a pair containing the
+ *  substring s and the const_iterator which points to c.
+ */
+template <typename ConstCharInputIterator, typename CharPredicate>
+std::pair<std::string, ConstCharInputIterator>
+takeWhile(ConstCharInputIterator Begin, ConstCharInputIterator End,
+          CharPredicate ShouldTake) {
+  std::stringstream WordStream;
+  auto Current = Begin;
+  while (Current != End && ShouldTake(*Current)) {
+    WordStream << *Current;
```

Comment:
I did this in f7bfbd17e1eb296fa822438bbc9bc7c46daa1fe8 but had to change it back again in f4a8b5b8e33a8c2193f78109ff5b7785e29d4b5d due to the surprising way that [std::istreambuf_iterator](http://en.cppreference.com/w/cpp/iterator/istreambuf_iterator) performs comparisons and the fact that it isn't a random access iterator.